### ISSUE TITLE
Run E2E Tests in CI

### ISSUE THREAD

**Issue by timvahlbrock** [Oeffis/Oeffis#171]

**Comment by timvahlbrock** [comment]
@Paulomart I could use some docker expertise on this. I was thinking about using the docker images we build in the CI to run the E2E tests. The problem is however, that buildx does not support loading multi-arch builds into the local registry at the moment: https://github.com/docker/buildx/issues/59 . I would say that the only option know is to scratch the idea of using the existing docker files and build some for E2E separately. Do you see any better options?

**Comment by Paulomart** [comment]
@timvahlbrock Maybe we cloud use this workaround? https://github.com/docker/roadmap/issues/371#issuecomment-1416932433

**Comment by timvahlbrock** [comment]
I'm giving it a shot. I don't like the idea of 2 builds either, especially because the message of "E2E pipeline has failed" might be confusing, if the cause actually are broken containers.

**Comment by timvahlbrock** [comment]
Setting provenance to false didn't help: https://github.com/Oeffis/Oeffis/actions/runs/5541249322/jobs/10114497978

**Comment by timvahlbrock** [comment]
The docker load approach seems to somewhat work, but for whatever reason it can't be loaded but only imported.

**Comment by timvahlbrock** [comment]
This seems to be what we want to do: https://docs.docker.com/build/ci/github-actions/share-image-jobs/

**Comment by timvahlbrock** [comment]
off-record communitcation between @Paulomart and me:

because upload artefacts is _really_ slow with (multi-arch) docker images, we decided to push the build containers to the registries with tags consisting of their commit hash and the attachment "-untested". The images can just be downloaded by the e2e job then and later be re-tagged as tested, given the testing succeeds.

**Comment by timvahlbrock** [comment]
<img width="1290" alt="image" src="https://github.com/Oeffis/Oeffis/assets/25803754/a17f69f9-e63f-4bf0-9241-5023ee010080">
@Paulomart any idea why the current frontend image build are getting stuck on npm ci in the builder?

**Comment by timvahlbrock** [comment]
Okay, I cleared the caches and am now running again with verbose logging enabled for npm. Seems like the npm package installation is just utterly slow currently.

**Comment by Paulomart** [comment]
@timvahlbrock Do you have a link for me to investigate?

**Comment by timvahlbrock** [comment]
some of the steps are now much faster, because the caches are filled again. but if the layer containing ci is build again, it still takes ages to complete.

**Comment by timvahlbrock** [comment]
what's strange is, https://github.com/Oeffis/Oeffis/actions/runs/5551531672/jobs/10137791995 did not use the cache, even though the package.json didn't change 🤔

**Comment by timvahlbrock** [comment]
It seems like, every time only one of the two images is build from cache. Maybe the cache keys are clashing?

**Comment by timvahlbrock** [comment]
Found: https://docs.docker.com/build/cache/backends/gha/#scope

Should be fixed now. Was actually right, cache scope by default is per branch, changed it to per branch per image.

**Comment by timvahlbrock** [comment]
Why didn't occur this earlier? Looks like we never actually build the frontend image in the CI, as its container building job was configured for the backend folder. So the caches also didn't clash.

**Comment by timvahlbrock** [comment]
@Paulomart maybe we can accelerate the container builds further if we make the builder layer shared between the platforms. Is this possible? Currently the arm64 building takes up a majority of the time required for building (20 minutes vs 10 seconds).

**Comment by Paulomart** [comment]
> @Paulomart maybe we can accelerate the container builds further if we make the builder layer shared between the platforms. Is this possible? Currently the arm64 building takes up a majority of the time required for building (20 minutes vs 10 seconds).

I think this is caused by the emulation of the cpu by the runner. Because different platforms use different base layers (as their platform differs) its not possible to share layers between them.

That said, an hacky way to speed up the builds, as we dont need any native compiled code, would be to build all files in an arch64 image and then copy the files into an prepared arm64 container.

**Comment by timvahlbrock** [comment]
> > @Paulomart maybe we can accelerate the container builds further if we make the builder layer shared between the platforms. Is this possible? Currently the arm64 building takes up a majority of the time required for building (20 minutes vs 10 seconds).
> 
> 
> 
> @timvahlbrock I think this is caused by the emulation of the cpu by the runner. Because different platforms use different base layers (as their platform differs) its not possible to share layers between them.
> 
> 
> 
> That said, an hacky way to speed up the builds, as we dont need any native compiled code, would be to build all files in an arch64 image and then copy the files into an prepared arm64 container. 

Okay. How I know you didn't like it much when we had it before, but couldn't we also use the build artifacts from the ci jobs and in the Docker stage just build them into the image?

**Comment by timvahlbrock** [comment]
Time spent: 6 hours

Spent on:

+ 4dc424020486ff8e89668ac5460fed399a4b1dad run oeffis containers in e2e pipeline job
+ be698f9f8137597e4160779a05955c1cf4491986 make e2e_test job depend on e2e_ci
+ 7b952837def5a6b7dc4d7de2e2881a538223b264 fix name of images in ci steps
+ 52fc4e1c8cdcaa4e7b440475313170418b2ede99 load container images in ci
+ 3b72ce4406d69834d09349d85e897dd47a09909f use custom containers for e2e
+ dfdee87081c957f6148b0ad866922aacf8e9867e don't load containers after building
+ fd10efd662792e8486d8f2a3807ab84c96870150 make container build/push dependent of e2e build
+ 8acd2da74f36d5301bbad2012d7d67c56563746c remove unnecessary upload image step
+ ac661e7d390709be60ac5bf32803cf281d3aa314 try to build with load, but without provenance
+ bbe422857b8aab84e3dbc66b54d59659a2e931c6 try to build container with oci output
+ c500d4fc98d853f581796fb551f2f1f59238fa1b try to load image from tar
+ a87ad5e0dbadbbf88e32f14309cf8cb9dc04796d set working directory when loading image
+ 06aac7fda77981b689db93ee9a59a1e477b94e6b disable tags in container image building
+ bba42fe10ba363ce2b74bdf8f602fb6dd085111b list backend dir
+ 410f98eef16c2b29bc57947870e5be548cc2f1b8 disable other ci jobs
+ ba3fc75dde3b2e991d8b26db62f7f52cc8df516d disable other ci jobs
+ 62d37e4eed900ccaaeee4de9a7ec554262411dfd try using docker import
+ 4c78b9c01fa39a4f909ebed1673ae589911f13b2 fix import usage
+ 5185cf4b07860e998445f9c276817bb69dd479bd try building/uploading image using instructions from docker documentation
+ cf9542d7516dd022cb6b28e818fd13238d83f561 switch to oci exporter again
+ 9c3a9c37e8d40ef53faa39e1c7c1c64ae055bf62 share images using docker registries
+ 9bb88c2132fb3fdc533cd8c97d17f10c67d05ad0 remove unnecessary (and breaking) escapes in tag names
+ abbfcd77facb1838d2510f3c7e76d7a9b94d3b07 push fix minor typos
+ ee43a6377975be75a3738bbf10b96a861b7346c7 add basic reachability test for actual app
+ 71fafdfaa8f1f728e4edb95b8ad19a37fefd104c align build step naming
+ 7c56a1fdd383ab9db4274679ff889b10952b52fd fix port mixup
+ fbaa540cca75e415659912cf27e0750b1afe023b revert previous port flip
+ 2f2bba12ded2adf9f76e373994d03572c1d91589 fix wrong network run command
+ a16d52a0c968410c8f29480dba722745c93c3c4c fix wrong context for frontend container
+ 0d3b0c5d9ea362f1c22c77de9cec3109ffe52224 launch containers as services
+ f182196f90b478da7f5d56f3e7d6b813d8ead561 clear all the caches
+ 35048cd73cb14d6de6dc62069c0558393a7d4d02 give cache clearing write permissions
+ d2505ddaaa2bce295cf0687f753fe580e7dcdd1a run cache clearing only
+ 33abd111fe400de46392b25cc3e549fc7b6dc2a5 reenable all steps
+ d4bac884b6affd1d5a5714e65de0cf6db040fd2c enable verbose logging for npm installs
+ 36a954091fcdcf4a3eba665e838919e0a8c84874 disable verbose logging for npm
+ 9d0dfaf78d9842fca756c4ee69c0d68637684a03 remove network option from ci configuration
+ d851d70019d49811037d9bad5a46729eae004a48 use cypress action to run tests
+ 698218c9481bb4d68d36967baaf7991498f98d76 always upload artifacts
+ 5f3944a1a760c94802156904c7eda3c2ee601d29 remove docker setup steps from e2e job
+ ce3e561e47c01d886956e420649d95b00c5b6955 only install arm64 qemu binaries
+ d6602e8368e123451583ba129ae192caee850270 fix artifact path
+ 7cda4c10515d99e06a693893944b780d31c0e8ff separate container image caches
+ 9d0ea4c13e748e83542ee1271efe27cc2dbebecd allow caching from base ref
+ a8aa9c6a5437a719b40728c54f4e4bc68716325c remove extra comma in cache list
+ 66c2d71a54603a139dcb7a262840025bb4503f4b fix port publishing
+ 168fccc7eb4d41bbfb6bf0f5265335a1e8ae93ee update upload artifact action to v3
+ 5581c8ee7689e084fd8ffad7c758615d23576329 add image publishing job
+ 06b1f7c2202108f03b40379705bd1d602f6e3b75 remove unnecessary job dependency on container images
+ 5671601f51fc6988fd9bc929a95660678a8f0280 add missing docker buildx
+ 62cee40fa40b18aa0aa3025119b213b7d9360da0 fix backend container contexts
+ 8f5c92d5a9fd7b99b891110f1faebc28e6e4153c enable pushing for tested images
+ b3ff75853712eb497f9dcf675334d7e3aaa4d512 restore swagger-spec.json
+ d89e8fe0f94a43ddca49575fb26ffbf24b0b6413 remove example duck duck go test

**Comment by timvahlbrock** [comment]
Regarding the performance problem:

I opened up a new issue for this: https://github.com/Oeffis/Oeffis/issues/180
